### PR TITLE
Abstracted Horovod params into the Backend API

### PR DESCRIPTION
**Pull request by tgaddair**
This PR is a precursor to adding a the Ray and Dask backends.

### CODE REVIEW

File: ludwig/collect.py [hunk 1]
```diff
@@ -321,11 +324,10 @@ def cli_collect_activations(sys_argv):
...
     parser.add_argument(
-        '-uh',
-        '--use_horovod',
-        action='store_true',
-        default=None,
-        help='uses horovod for distributed training'
+        "-b",
+        "--backend",
+        help="specifies backend to use for parallel / distributed execution",
+        choices=ALL_BACKENDS,
```
Comment (w4nderlust) [inline ludwig/collect.py:330]: I guess the default is None if we don't specify it, but maybe being explicit that the default is the local one may be easier for the users and for the help message?

File: ludwig/evaluate.py [hunk 2]
```diff
@@ -242,11 +243,10 @@ def cli(sys_argv):
...
     parser.add_argument(
-        '-uh',
-        '--use_horovod',
-        action='store_true',
-        default=None,
-        help='uses horovod for distributed training'
+        "-b",
+        "--backend",
+        help="specifies backend to use for parallel / distributed execution",
+        choices=ALL_BACKENDS,
```
Comment (w4nderlust) [inline ludwig/evaluate.py:249]: same as above

File: ludwig/experiment.py [hunk 3]
```diff
@@ -532,11 +533,10 @@ def cli(sys_argv):
...
     parser.add_argument(
-        '-uh',
-        '--use_horovod',
-        action='store_true',
-        default=None,
-        help='uses horovod for distributed training'
+        "-b",
+        "--backend",
+        help="specifies backend to use for parallel / distributed execution",
+        choices=ALL_BACKENDS,
```
Comment (w4nderlust) [inline ludwig/experiment.py:539]: same as above

File: ludwig/features/category_feature.py [hunk 4]
```diff
@@ -394,12 +393,7 @@ def postprocess_predictions(
         postprocessed = {}
         name = self.feature_name
 
-        npy_filename = None
-        if is_on_master():
```
Comment (w4nderlust) [inline ludwig/features/category_feature.py:398]: here the logic was that each worker was actually doing postprocessing so they would override each other's outputs if they were on the same machine (like one machine with multiple gpus). I'm not sure about how this issue is addressed now.

Comment (w4nderlust) [reply ludwig/features/category_feature.py:398]: Replicated for all output features

File: ludwig/hyperopt_cli.py [hunk 5]
```diff
@@ -365,11 +366,10 @@ def cli(sys_argv):
...
     parser.add_argument(
-        "-uh",
-        "--use_horovod",
-        action="store_true",
-        default=False,
-        help="uses horovod for distributed training",
+        "-b",
+        "--backend",
+        help="specifies backend to use for parallel / distributed execution",
+        choices=ALL_BACKENDS,
```
Comment (w4nderlust) [inline ludwig/hyperopt_cli.py:372]: same as above

File: ludwig/models/ecd.py [hunk 6]
```diff
@@ -271,6 +278,14 @@ def recurse_weights(model, prefix=None):
 
         return weights
 
+    def __setstate__(self, newstate):
```
Comment (w4nderlust) [inline ludwig/models/ecd.py:281]: trying to get a better understanding of __setstate__ and __reduce__. Why are they needed? How are they used?

File: ludwig/models/predictor.py [hunk 7]
```diff
@@ -10,17 +11,61 @@
...
+            dataset,
+            bucketing_field=None
+    ):
+        raise NotImplementedError()
+
+    # Remote implementations may override this
+    def shutdown(self):
+        pass
+
+    # Functions needed to treat Trainer as a context manager
```
Comment (w4nderlust) [inline ludwig/models/predictor.py:60]: Trainer -> Predictor

File: ludwig/models/trainer.py [hunk 8]
```diff
@@ -191,8 +236,8 @@ def __init__(
         self.batch_size = batch_size
         self.eval_batch_size = batch_size if eval_batch_size < 1 else eval_batch_size
         self.bucketing_field = bucketing_field
-        self.validation_field = validation_field
-        self.validation_metric = validation_metric
+        self._validation_field = validation_field
```
Comment (w4nderlust) [inline ludwig/models/trainer.py:239]: why are _ added here?

File: ludwig/predict.py [hunk 9]
```diff
@@ -216,11 +217,10 @@ def cli(sys_argv):
...
     parser.add_argument(
-        '-uh',
-        '--use_horovod',
-        action='store_true',
-        default=None,
-        help='uses horovod for distributed training'
+        "-b",
+        "--backend",
+        help="specifies backend to use for parallel / distributed execution",
+        choices=ALL_BACKENDS,
```
Comment (w4nderlust) [inline ludwig/predict.py:223]: same as above

File: ludwig/preprocess.py [hunk 10]
```diff
@@ -238,6 +240,12 @@ def cli(sys_argv):
              'to a random number generator: data splitting, parameter '
              'initialization and training set shuffling'
     )
+    parser.add_argument(
+        "-b",
+        "--backend",
+        help="specifies backend to use for parallel / distributed execution",
+        choices=ALL_BACKENDS,
```
Comment (w4nderlust) [inline ludwig/preprocess.py:247]: same as above

File: ludwig/train.py [hunk 11]
```diff
@@ -375,11 +376,10 @@ def cli(sys_argv):
...
     parser.add_argument(
-        '-uh',
-        '--use_horovod',
-        action='store_true',
-        default=None,
-        help='uses horovod for distributed training'
+        "-b",
+        "--backend",
+        help="specifies backend to use for parallel / distributed execution",
+        choices=ALL_BACKENDS,
```
Comment (w4nderlust) [inline ludwig/train.py:382]: same as above

File: requirements.txt [hunk 12]
```diff
@@ -12,6 +12,7 @@ PyYAML>=3.12
 absl-py
 kaggle
 requests
+tables
```
Comment (w4nderlust) [inline requirements.txt:15]: why is this required?

File: ludwig/features/category_feature.py (hunk 4 above)
Comment (tgaddair) [reply ludwig/features/category_feature.py:398]: This is handled in the `postprocess` function which calls into these output feature functions here: https://github.com/ludwig-ai/ludwig/blob/backend_refactor/ludwig/data/postprocessing.py#L35

This saves us from having to pass around the `backend` to all these features just for checking this one condition, and removes the duplicate code.

File: ludwig/collect.py (hunk 1 above)
Comment (tgaddair) [reply ludwig/collect.py:330]: It's usually the local backend, but the tricky thing is that if the user runs with Horovod like: `horovodrun ludwig train ...` then the backend will be Horovod.  We can specify this in the help message if you think it would add clarity, but I think we should leave the value itself as `None` for this reason.

File: ludwig/models/trainer.py (hunk 8 above)
Comment (tgaddair) [reply ludwig/models/trainer.py:239]: These needed to be made private so that we could hide the public interface behind a property here:

https://github.com/ludwig-ai/ludwig/blob/backend_refactor/ludwig/models/trainer.py#L819

This is important for the upcoming Ray backend, which needs to override this behavior.

File: requirements.txt (hunk 12 above)
Comment (tgaddair) [reply requirements.txt:15]: This is actually needed for the Dask PR, if I remember correctly, so I'll remove this for now and add it in that PR.

File: ludwig/collect.py (hunk 1 above)
Comment (w4nderlust) [reply ludwig/collect.py:331]: Oh you are right, forgot about that. Yes specifying in the message that horovodrun makes it autodetect horovod, and that this overrides the autodetection is a good idea for clarity.

File: requirements.txt (hunk 12 above)
Comment (w4nderlust) [reply requirements.txt]: not a huge deal, but moving it will likely keep things cleaner

Comment (tgaddair) [reply requirements.txt]: Fixed.
